pkg/server/plugin/keymanager/hashicorpvault/hashicorp_vault.go

```diff
@@ -0,0 +1,490 @@
+
+	logger  hclog.Logger
+	mu      sync.RWMutex
+	entries map[string]keyEntry
+
+	authMethod AuthMethod
+	cc         *ClientConfig
+	vc         *Client
+
+	hooks pluginHooks
+}
+
+// New returns an instantiated plugin.
+func New() *Plugin {
+	return newPlugin()
+}
+
+// newPlugin returns a new plugin instance.
+func newPlugin() *Plugin {
+	return &Plugin{
+		entries: make(map[string]keyEntry),
+		hooks: pluginHooks{
+			lookupEnv: os.LookupEnv,
+		},
+	}
+}
+
+// SetLogger sets a logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.logger = log
+}
+
+func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	config := new(Config)
+
+	if err := hcl.Decode(&config, req.HclConfiguration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	am, err := parseAuthMethod(config)
+	if err != nil {
+		return nil, err
+	}
+	cp, err := p.genClientParams(am, config)
+	if err != nil {
+		return nil, err
+	}
+	vcConfig, err := NewClientConfig(cp, p.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	p.authMethod = am
+	p.cc = vcConfig
+
+	if p.vc == nil {
+		err := p.genVaultClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	p.logger.Debug("Fetching keys from Vault")
+	keyEntries, err := p.vc.GetKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	p.setCache(keyEntries)
+
+	return &configv1.ConfigureResponse{}, nil
+}
+
+func parseAuthMethod(config *Config) (AuthMethod, error) {
+	var authMethod AuthMethod
+	if config.TokenAuth != nil {
+		authMethod = TOKEN
+	}
+
+	if config.AppRoleAuth != nil {
+		if err := checkForAuthMethodConfigured(authMethod); err != nil {
+			return 0, err
+		}
+		authMethod = APPROLE
+	}
+
+	if config.CertAuth != nil {
+		if err := checkForAuthMethodConfigured(authMethod); err != nil {
+			return 0, err
+		}
+		authMethod = CERT
+	}
+
+	if config.K8sAuth != nil {
+		if err := checkForAuthMethodConfigured(authMethod); err != nil {
+			return 0, err
+		}
+		authMethod = K8S
+	}
+
+	if authMethod != 0 {
+		return authMethod, nil
+	}
+
+	return 0, status.Error(codes.InvalidArgument, "one of the available authentication methods must be configured: 'Token, AppRole'")
+}
+
+func checkForAuthMethodConfigured(authMethod AuthMethod) error {
+	if authMethod != 0 {
+		return status.Error(codes.InvalidArgument, "only one authentication method can be configured")
+	}
+	return nil
+}
+
+func (p *Plugin) genClientParams(method AuthMethod, config *Config) (*ClientParams, error) {
+	cp := &ClientParams{
+		VaultAddr:         p.getEnvOrDefault(envVaultAddr, config.VaultAddr),
+		Namespace:         p.getEnvOrDefault(envVaultNamespace, config.Namespace),
+		TransitEnginePath: p.getEnvOrDefault(envVaultTransitEnginePath, config.TransitEnginePath),
+		CACertPath:        p.getEnvOrDefault(envVaultCACert, config.CACertPath),
+		TLSSKipVerify:     config.InsecureSkipVerify,
```

Comment:
I think that we should log a warning if InsecureSkipVerify is set to true.

Reply:
Added a warning and adjusted the docs